xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/hibernate/internal/IDVersionHibernateAdapterFactory.java

```diff
@@ -0,0 +1,168 @@
+    private List<DatabaseProductNameResolver> resolvers;
+
+    @Override
+    public Optional<HibernateAdapter> createHibernateAdapter(DatabaseMetaData metaData,
+        Configuration hibernateConfiguration) throws HibernateException
+    {
+        HibernateAdapter adapter = null;
+
+        // Index all adapters
+        Map<String, Map<Version, String>> mapping = getAdapterMapping();
+
+        // Gather adapter(s) associated to the specific database product id
+        Map<Version, String> versions;
+        try {
+            // Resolved the database product identifier
+            String databaseProductId = getDatabaseId(metaData.getDatabaseProductName());
+
+            versions = mapping.get(databaseProductId);
+        } catch (SQLException e) {
+            throw new HibernateException("Failed to access the database product name", e);
+        }
+
+        if (versions != null) {
+            // Resolve the current database version
+            Version currentVersion;
+            try {
+                currentVersion = new DefaultVersion(metaData.getDatabaseProductVersion());
+            } catch (SQLException e) {
+                throw new HibernateException("Failed to access the database product version", e);
+            }
+
+            // Search if a registered adapter matched the current version
+            String roleHint = null;
+            for (Map.Entry<Version, String> entry : versions.entrySet()) {
+                if (entry.getKey().compareTo(currentVersion) <= 0) {
+                    roleHint = entry.getValue();
+                } else {
+                    break;
+                }
+            }
+
+            // Load the found adapter
+            if (roleHint != null) {
+                try {
+                    adapter = this.componentManager.getInstance(HibernateAdapter.class, roleHint);
+                } catch (ComponentLookupException e) {
+                    throw new HibernateException("Failed to initialize the adapater", e);
+                }
+            }
+        }
+
+        return Optional.ofNullable(adapter);
+    }
+
+    private String getDatabaseId(String databaseProductName)
+    {
+        for (DatabaseProductNameResolver resolver : this.resolvers) {
+            Optional<String> id = resolver.resolve(databaseProductName);
+
+            if (id.isPresent()) {
+                return id.get();
+            }
+        }
+
+        return databaseProductName.toLowerCase();
+    }
+
+    private Map<String, Map<Version, String>> getAdapterMapping()
+    {
+        List<ComponentDescriptor<HibernateAdapter>> descriptors =
+            this.componentManager.getComponentDescriptorList(HibernateAdapter.class);
+
+        Map<String, Map<Version, String>> databases = new HashMap<>();
+        for (ComponentDescriptor<HibernateAdapter> descriptor : descriptors) {
+            String roleHint = descriptor.getRoleHint();
+            if (roleHint != null && !"default".equalsIgnoreCase(roleHint)) {
```

Comment:
we don't have a constant for "default" in ComponentDescriptor?

Reply:
I guess a `ComponentDescriptor#isDefaultRoleHint(String)` would be nice.